### ISSUE TITLE
AWS Secrets Manager Extension for DB Config management  

### ISSUE THREAD

**Issue by gboro54** [quarkusio/quarkus#6896]
Similar to the Vault extension, we should consider adding an extenion that can communicate to AWS Secrets Manager and override DB Connection information (username and passwords) at the time the connection pool is being created in the AbstractDataSourceProducer class. 

This would intel perhaps making the CredentialsProvider interface more generic (i.e outside of the the Vault extension) to allow additional providers to leverage the same interface for injection of these properties. 

**Configuration suggestion**
quarkus.secretsmanager.aws.region 
quarkus.secretsmanager.aws.credentials.type
quarkus.secretsmanager.aws.secretsid

**Additional context**
Another thought would be to make the vault extension a bit more generic, however this does not seem to be in the spirt of Quarkus from a plugin/extension perspective

**Comment by sberyozkin** [comment]
@gboro54 Vault extension has been done by Vincent @vsevel such that the alternative `Vault` implementations can be plugged in. However in this case it appears Vault has an [AWS secrets engine](https://www.vaultproject.io/docs/secrets/aws/). 
Hi Vincent, I've signed to you for you to review if we should go ahead with the Vault AWS engine
thanks

**Comment by sberyozkin** [comment]
No it looks different really to what @gboro54 is asking about. @vsevel, I suppose one should go using the CredentialsProvider (vault-spi) route if an alternative Vault is required ? thanks

**Comment by sberyozkin** [comment]
@gboro54 `quarkus-agroal` has only [this link](https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalVaultCredentialsProviderPassword.java) to a `Vault` and you can see how it is linked at the configuration level [here](https://github.com/quarkusio/quarkus/blob/master/integration-tests/vault-agroal/src/test/resources/application-vault-datasource.properties).
So, if you'd like you can introduce an AWS secrets managed extension, which would depend on `quarkus-vault-spi` only and provide it is own `CredentialsProvider` - please experiment

**Comment by gboro54** [comment]
@sberyozkin - That is my thought. I am currently doing this in my component but am going to work on cleaning it up and moving it out into its own module. I will take a look at docs around extensions to add all the needed pieces and see if I can get a PR in the next week.

**Comment by vsevel** [comment]
@sberyozkin you are right in assuming that the credentials provider had been designed with that kind of extended use cases in mind. however the aws secret engine in hashicorp vault seems to be related to authentication, rather than secret mgmt.

@gboro54 here what you would do to integrate an alternate vault:

create an extension that has all the logic to connect to the aws vault, including config elements that are specific to aws. here is what has been done for vault: [vault runtime config](https://github.com/quarkusio/quarkus/blob/master/extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/VaultRuntimeConfig.java#L180-L191), and [CredentialsProviderConfig](https://github.com/quarkusio/quarkus/blob/master/extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/config/CredentialsProviderConfig.java) (for the vault specific properties that are relevant to configuring a credentials provider).

implement the `CredentialsProvider` interface. for instance: [VaultCredentialsProvider](https://github.com/quarkusio/quarkus/blob/master/extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/VaultCredentialsProvider.java)

the `CredentialsProvider` interface's contract is simply to return a `Properties` object, which will be used as an agroal [SimplePassword](https://github.com/agroal/agroal/blob/master/agroal-api/src/main/java/io/agroal/api/security/SimplePassword.java) through [AgroalVaultCredentialsProviderPassword](https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalVaultCredentialsProviderPassword.java).

ultimately this will end up in the [AgroalDefaultSecurityProvider](https://github.com/agroal/agroal/blob/master/agroal-api/src/main/java/io/agroal/api/security/AgroalDefaultSecurityProvider.java#L21-L23). so whatever properties get returned by the credentials provider, they need to be understood by the agroal [ConnectionFactory](https://github.com/agroal/agroal/blob/master/agroal-pool/src/main/java/io/agroal/pool/ConnectionFactory.java#L197-L210).

in the end your config might look like this, assuming your extension is called `awsvault`:
```
quarkus.awsvault.credentials-provider.myawsvault.region=europe
quarkus.awsvault.credentials-provider.myawsvault.credentials.type=XXX
quarkus.awsvault.credentials-provider.myawsvault.secretsid=XXX

quarkus.datasource.driver=org.postgresql.Driver
quarkus.datasource.url=jdbc:postgresql://localhost:6543/mydb
quarkus.datasource.username=postgres
quarkus.datasource.credentials-provider=myawsvault
```

tell me how it goes

**Comment by gboro54** [comment]
@vsevel - Thanks for the info. I am working on the details now.

**Comment by vsevel** [comment]
documentation now provided in #9552

**Comment by yahmed153** [comment]
Hi all, any idea when an AWS CredentialProvider might be completed. Or should I go ahead and roll my own. Thanks for your time.

**Comment by sberyozkin** [comment]
@yahmed153 Can you look at #10663 please ?

**Comment by yahmed153** [comment]
> @yahmed153 Can you look at #10663 please ?

Will do thanks

**Comment by psaraf** [comment]
is this still under discuss, any plan when it will be released. really like the approach of configuration.

**Comment by vsevel** [comment]
FYI I created a POC for the Azure Key Vault: https://github.com/vsevel/poc-akv
this is providing a config source (as opposed to the credentials provider approach).

it should be easy to adapt for aws; the main logic is in:
- [AzureKeyVaultConfigSource](https://github.com/vsevel/poc-akv/blob/master/runtime/src/main/java/io/quarkiverse/quarkus/azure/keyvault/runtime/config/AzureKeyVaultConfigSource.java)
- [AzureKeyVaultBootstrapConfig](https://github.com/vsevel/poc-akv/blob/master/runtime/src/main/java/io/quarkiverse/quarkus/azure/keyvault/runtime/config/AzureKeyVaultBootstrapConfig.java)
- [AzureKeyVaultVertxClient](https://github.com/vsevel/poc-akv/blob/master/runtime/src/main/java/io/quarkiverse/quarkus/azure/keyvault/runtime/client/AzureKeyVaultVertxClient.java)

**Comment by temiotu** [comment]
is this still under discussion, any plan when it will be released??
